kep: dynamic plugin configuration (#2807)
Kong/kubernetes-ingress-controller: keps/0010-dynamic-plugin-configuration.md

difficult to manage. While it is possible to load configuration from a secret using
the `configFrom` parameter, the entire configuration must be contained with in the
secret which is effectively the same problem.

It would be advantageous, and more in line with common Kubernetes behavior to be able
to derive individual values of plugin configuration from data already stored on the same
Kubernetes cluster. This proposal aims to define a change to the `KongPlugin` and `KongClusterPlugin`
Custom Resource Definitions to facilitate this behavior.

## Motivation

Allow `KongPlugin` and `KongClusterPlugin` resources to read configuration values from Kubernetes,
in addition to literal values.

Continue to allow `KongPlugin` and `KongClusterPlugin` resource to accept literal input values.

Reduce the amount of duplication of configuration value in the simple case. That is,
when using standard kubernetes objects and with out assumed templating systems.

### Goals

* Provide ability to read individual configuration values from `ConfigMap` resources
* Provide ability to read individual configuration values from `Secret` resources

> * Provide ability to read individual configuration values from arbitrary or generic resource types (other custom resource types)

Reviewer: Did you have anything specific in mind here? If not, I would ask that we keep things tight by sticking with the use cases we know. If so, perhaps there's a user-story we can add that will help support this goal?

Author: I suppose this really give me the ability to read data from any other CRD that may store data

```yaml
    - genericMapKeyRef:
        apiVersion: a.custom.resource/v1
        kind: ResourceFoo
        name: foo-bar
        namespace: default
```

with this sort of an implementation, doing configmap + secret key look ups aren't that much different and keep the internals consistent + simple

Reviewer: This would seem to be a source of significant complexity. For the moment since it would appear this may be mostly a "want" rather than a "need" do you think it would be reasonable to add this to the alternatives considered section as something possibly for a future iteration?

Author: of course. secrets + configmaps are certainly the need.

I bring it up as I've implement this in [another controller](https://github.com/razee-io/razeedeploy-core/blob/86981ef2ccb61526d9bcb3848fa39f89c928a0b6/lib/FetchEnvs.js#L50-L64)  and it works out rather cleanly.

### Non-Goals

> * Introduction of a new Custom Resource or different Plugin Resource

Reviewer: This reads to me that you're trying to avoid is new APIs, and more specifically you would like to be able to convert existing `KongPlugins` in a backwards compatible manner (e.g. conversion webhook) to any new APIs. Is that a correct interpretation?

Author: Yes, that the intention here. @rainest indicated that there was a desire to do that and maintain backward and forward compatibility. Which I can apprieciate. Migrating off of deprecated CRDs isn't always easy.

If there is an entirely new resource type - not a deal breaker to me

Reviewer: Ok, let me know what you think about this language change:

```suggestion
* We want to avoid introducing any new resource kinds, instead keeping with `KongPlugin`
```

Author: Works for me

## Proposal

In an effort to stay in line with current Kubernetes resource patterns it would be desirable
to mimic the way pods define environment variables using an array of objects where
the objects contain a `name` and one of `value` or `valueFrom`. `valueFrom`

> would further supports `configMapKeyRef`, `secretKeyRef` or `genericKeyRef`

Reviewer: :+1: , but I almost feel like this should be listed among the goals.

Author: Can do

### User Stories (Optional)

#### Story 1

As a operator I want certain plugin configuration values to be dynamic in addition
to statically definable

#### Story 2

As a operator I want to be able to derive a literal value using value interpolation referencing
other values

#### Story 3

> As an operator I want values stored in `ConfigMap` or `Secret` resources
> to be the source of truth for various resources deployed to my Kubernetes cluster

Reviewer: I'm not sure I'm understanding this user story at a glance, is this paraphrase a good interpretation?

```suggestion
As an operator I want to be able to store common plugin options and configurations as
`ConfigMaps` or `Secrets` such that I can write common configurations once and not
have to re-write them across multiple plugins.
```

Author: Yes. I suppose I was trying to indicate that we want to be able to leverage configuration that already exists on the cluster and that isn't put there specifically for plugins. re-use across resources on a cluster.

But, yes this captures it in the context of what the controller cares about.

#### Story 4

As an operator I expect changes in `ConfigMap` or `Secret` data values to be reflected

> in the resource and deployments that depend on them with out having to restart critical parts of the system

Reviewer: This is a particularly interesting user story. It would seem we would want to add something to the goals for this. Perhaps something like:

- the `KongPlugin` controller should react to updates to `ConfigMap` and `Secret` objects referenced by managed plugins to re-apply the configuration to all attached `KongPlugins`

Does that resonate with you?

Also this has some implications in terms of danger around operators changing commonly used `ConfigMaps` and `Secrets` as those changes could cascade to any number of otherwise configured and operating plugins. Any thoughts on some of the potential hazards?

Author: Right. one of the very nice things about the plugins now being basically static + the controller loops is they are effectively Hot reloadable. updating the plugin - wait 5 seconds.

In this model that breaks, and I would have to restart the kong deployments - which has implications.

Stress on the kubernetes API server is a real thing. we use a controller the just runs on an interval and that has been far worse than event driven updates. Another problem is going to be resources that get deleted may have a negative cascade / deck of cards effect. You might be able to prevent that with a finalizer + child resources (maybe?). 

Its probably not save to assume everyone does this, but we keep config maps pretty tight to use case - all redis config is in 1  `redis` for example. Try to avoid dumping ground config map that would have high frequency updates.

As an alternative, I've seen other controllers use annotations to allow opt-in to watching other resources. not sure I'd like that in this case.

Author: I don't think the cascade of changing 1 secret down to many plugins is a bad thing, I'd expect that.

Reviewer: > I don't think the cascade of changing 1 secret down to many plugins is a bad thing, I'd expect that.

Indeed. I certainly agree it's an obvious implication, but I was more wondering if you feel like there's anything we can do to ease the pain of unintended consequence: e.g. if someone were to update a `ConfigMap` with an accidental _bad configuration_ (think similar to when ISPs accidentally mess up their BGP configurations, but certainly on a much smaller scale) are there any things we can do to reduce or help in these situations :thinking: I'm kinda thinking it's a _no_, and we just have to consider the risk part of the cost of offering this kind of solution.

On the other hand, for the purposes of resolving this comment I would just like to see us reflect this user story as a goal, perhaps similar to what I suggested above, as it seems this is a fairly important design goal.

Author: well there is some validation at the plugin level with the admission controller which should prevent things from being overtly invalid. outside of that, I think that is mostly contextual - not much you can control at that point

Author: Added to goals as well

#### Story 5

As an operator I expect optional dynamic values to support default values to be utilized

> when the key or resource is missing from the cluster

Reviewer: I'm sorry to say it's not clear to me what this story intends. Is this suggesting that when someone makes a reference to a `ConfigMap` or a `Secret` but that reference can't be fufilled we have backup configuration options which try to fill in the missing pieces? I could see this as being potentially quite difficult and maybe hazardous if we try to do it in an automagical kind of way :thinking:

Author: Yes, that.
But, not magic, explicit.

```yaml
configFrom:
  - name: FOO
    valueFrom:
      configMapKeyRef:
        name: my-secret
        key: foo
        optional: true
        default: 100   # <-----  a sane default
        type: number
```

Reviewer: Ah I see that example makes perfect sense, thank you. Perhaps this wording could help:

```suggestion
As an operator I expect to have the option to provide default values for plugin configuration which
will be used in the case that a referent object is not available or readable.
```

Let me know your thoughts? :thinking:

Author: I like it. Done

### Notes/Constraints/Caveats (Optional)

Using the data defined from the existing `config` values first and utilizing the
`configFrom` values may provide a simple transition forward. This would allow existing
configuration to continue to work as it currently does and values defined by `configFrom`

> override. Combined with a feature gate, it may provide means to move between styles

Reviewer: An alternative to this we should consider is making a `v2` of the `KongPlugin` API, with a conversion webhook that can convert between them.

Author: That was my original thought / suggestion on #2763. @rainest indicated you all try to maintain backward compatibility, and that option would make that difficult. 

If it isn't the case - I'm very in favor of this option

Reviewer: Sounds good, can we capture that perhaps in the notes or a TODO list (or have we already and I just have missed it)? I would consider this comment resolved as long as we have some kind of follow up on that for posterity.

Author: added as an alternative. And as additions to `notes`

        type: number
```

Pods also allow for previously read values to be interpolated to form new literal values.
This behavior is desirable

**Value Interpolation**

```yaml
apiVersion: configuration.konghq.com/v1
kind: KongPlugin
metadata:
  name: plugin-one
config:
  bar: hello-world
configFrom:
  - name: DB_USERNAME
    valueFrom:
      secretMapKeyRef:
        name: my-secret
        key: username

  - name: DB_PASSWORD
    valueFrom:
      secretMapKeyRef:
        name: my-secret
        key: password

  - name: DB_OPTIONS
    valueFrom:
      secretMapKeyRef:
        name: my-config-map
        key: options

  - name: DB_URL
    value: db://$(DB_USERNAME):$(DB_PASSWORD)@dbhost?$(DB_OPTIONS)
```

#### Caveats

TODO

## Drawbacks

* Support for backwards compatibility between the current implementation, and this style
may not be entirely possible.
* Supporting both the existing literal syntax and the proposed syntax at the same time may be too difficult to maintain

## Alternatives

* It is additionally possible to maintain the existing `config` and `configFrom` keys as they exist
and introduce an additional key that supports the behaviors outlines in this proposal

> 

Reviewer: ```suggestion
```
